File: python/techai_webutils/tests/unit/clients_client_stack_test.py
"""Behavioural tests for the async client resilience stack (new_client_stack_from_config).

Why this suite matters:
    - The direct-SDK async clients (Bedrock, Ollama, S3, SQS) get their timeout / retry /
      circuit-breaker / bulkhead protection from this composed stack; each layer must behave
      correctly and the disabled config must yield a bare passthrough.
"""

from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from unittest.mock import AsyncMock, MagicMock, create_autospec

import pytest

from techai_webutils.clients.decorators.proxy import (
    ClientStackConfig,
    RateLimitProxy,
    new_client_stack_from_config,
)
from techai_webutils.core.interfaces.rate_limiter import RateLimiter
from techai_webutils.core.errors.errors import AppError, ErrorCode
from techai_webutils.foundation.resilience.circuit_breaker import (
    CircuitBreaker,
    CircuitOpenError,
)
from techai_webutils.foundation.resilience.classify import is_transient


def _client(**do_config: object) -> MagicMock:
    """A mock client whose async ``do`` is an ``AsyncMock`` configured with ``do_config``.

    The resilience proxies forward ``getattr(client, "do")`` and branch on
    ``inspect.iscoroutinefunction`` (True for ``AsyncMock``), so this stands in for the direct-SDK
    client the stack wraps. ``do.await_count`` gives the invocation count the old fake tracked by hand.
    """
    client = MagicMock()
    client.do = AsyncMock(**do_config)
    return client


def _sleeping_do(seconds: float) -> Callable[..., Awaitable[str]]:
    """An async ``do`` side effect that sleeps ``seconds`` then returns ``"ok"``."""

    async def _do(*_args: object, **_kwargs: object) -> str:
        await asyncio.sleep(seconds)
        return "ok"

    return _do


@pytest.mark.asyncio
async def test_client_stack_disabled_config_is_passthrough() -> None:
    """A disabled config returns the bare client (backwards-compatible path).

    What it tests: the composer returns the wrapped object itself and calling it invokes the
    method directly with no resilience layers.
    """
    fake = _client(return_value="ok")
    stacked = new_client_stack_from_config(fake, "t", ClientStackConfig(enabled=False))
    assert stacked is fake
    assert await stacked.do() == "ok"


@pytest.mark.asyncio
async def test_client_stack_retries_transient_then_succeeds() -> None:
    """A Retryable transient failure is retried by the stack's Retry layer.

    What it tests: with retry enabled, a call failing once transiently then succeeding is
    invoked twice and returns success.
    """
    fake = _client(side_effect=[AppError(ErrorCode.UNAVAILABLE, "transient"), "ok"])
    cfg = ClientStackConfig(
        retry_enabled=True, retry_max_attempts=3, bulkhead_max_concurrent=None, timeout_seconds=None
    )
    stacked = new_client_stack_from_config(fake, "t", cfg)
    assert await stacked.do() == "ok"
    assert fake.do.await_count == 2


@pytest.mark.asyncio
async def test_client_stack_circuit_opens_after_threshold() -> None:
    """The injected circuit breaker trips after repeated failures and then fails fast.

    What it tests: once the breaker opens, a call raises CircuitOpenError without invoking the
    wrapped method.
    """
    fake = _client(side_effect=AppError(ErrorCode.UNAVAILABLE, "transient"))  # always fails
    cb = CircuitBreaker(failure_threshold=2)
    cfg = ClientStackConfig(bulkhead_max_concurrent=None, timeout_seconds=None)
    stacked = new_client_stack_from_config(fake, "t", cfg, circuit_breaker=cb)

    open_hit = False
    for _ in range(6):
        try:
            await stacked.do()
        except CircuitOpenError:
            open_hit = True
            break
        except AppError:
            pass
    assert open_hit, "breaker should open after repeated failures"

    calls_at_open = fake.do.await_count
    with pytest.raises(CircuitOpenError):
        await stacked.do()
    assert fake.do.await_count == calls_at_open, "open breaker must not invoke the wrapped method"


@pytest.mark.asyncio
async def test_client_stack_bulkhead_limits_concurrency() -> None:
    """The bulkhead bounds concurrent in-flight calls.

    What it tests: with max_concurrent=1, two concurrent calls never overlap (peak in-flight 1).
    """
    in_flight = 0
    peak = 0

    async def _tracking_do(*_args: object, **_kwargs: object) -> str:
        """Record peak concurrency while sleeping, so the bulkhead's serialization is observable."""
        nonlocal in_flight, peak
        in_flight += 1
        peak = max(peak, in_flight)
        try:
            await asyncio.sleep(0.05)
            return "ok"
        finally:
            in_flight -= 1

    fake = _client(side_effect=_tracking_do)
    cfg = ClientStackConfig(bulkhead_max_concurrent=1, timeout_seconds=None)
    stacked = new_client_stack_from_config(fake, "t", cfg)
    await asyncio.gather(stacked.do(), stacked.do())
    assert peak == 1


@pytest.mark.asyncio
async def test_client_stack_preserves_configured_timeout() -> None:
    """The Timeout layer applies the configured deadline verbatim, never a shorter default.

    What it tests: a fast call under a long (300s) timeout succeeds (not shortened), while a slow
    call under a short timeout is cut off.
    """
    fast = _client(side_effect=_sleeping_do(0.01))
    fast_stacked = new_client_stack_from_config(
        fast, "t", ClientStackConfig(timeout_seconds=300.0, bulkhead_max_concurrent=None)
    )
    assert await fast_stacked.do() == "ok"

    slow = _client(side_effect=_sleeping_do(0.2))
    slow_stacked = new_client_stack_from_config(
        slow, "t", ClientStackConfig(timeout_seconds=0.05, bulkhead_max_concurrent=None)
    )
    with pytest.raises((TimeoutError, AppError)):
        await slow_stacked.do()


class _Permanent:
    """Async fake that always raises a PERMANENT (non-transient) domain error; counts calls."""

    def __init__(self) -> None:
        """Start the call counter at zero."""
        self.calls = 0

    async def do(self) -> str:
        """Increment the counter and raise a permanent NotFound error."""
        self.calls += 1
        raise AppError(ErrorCode.NOT_FOUND, "gone")


@pytest.mark.asyncio
async def test_client_stack_permanent_error_does_not_trip_breaker() -> None:
    """A permanent domain error must NOT trip the circuit breaker (only transient infra failures do).

    Why this test is important:
        - A breaker that trips on ordinary business outcomes (NotFound, validation) fails healthy
          callers; the breaker must be error-aware, mirroring the Go gobreaker IsSuccessful (audit R1).

    What it tests:
        - With a classifying breaker (is_failure=is_transient), repeated permanent AppErrors never open
          the circuit, so every call still reaches the wrapped method.
    """
    fake = _Permanent()
    cb = CircuitBreaker(failure_threshold=2, is_failure=is_transient)
    cfg = ClientStackConfig(bulkhead_max_concurrent=None, timeout_seconds=None)
    stacked = new_client_stack_from_config(fake, "t", cfg, circuit_breaker=cb)

    for _ in range(6):
        with pytest.raises(AppError):
            await stacked.do()
    assert fake.calls == 6, "permanent errors must not trip the breaker"


def test_client_stack_wrap_order_tracing_outermost_logging_innermost() -> None:
    """The composed client stack nests Tracing -> Logging (Logging innermost), per charter §6.3.

    Why this test is important:
        - The tracing span must bracket the logging seam; before the Go builders had
          this inverted (Logging outermost). This locks the Python composer's order so a reorder
          of the LoggingProxy/TracingProxy composition fails the test.

    What it tests:
        - With only logging + tracing enabled, the outermost proxy is a TracingProxy wrapping a
          LoggingProxy wrapping the target — so Logging is the innermost seam.
    """

    class _Target:
        async def do(self) -> str:
            return "ok"

    target = _Target()
    stacked = new_client_stack_from_config(
        target,
        "t",
        ClientStackConfig(timeout_seconds=None, retry_enabled=False, bulkhead_max_concurrent=None),
    )

    assert type(stacked).__name__ == "TracingProxy", "Tracing must be outermost of the trio"
    inner = stacked._wrapped  # noqa: SLF001 — structural assertion of the composed nesting
    assert type(inner).__name__ == "LoggingProxy", "Logging must sit inside Tracing (innermost)"
    assert inner._wrapped is target  # noqa: SLF001 — LoggingProxy wraps the target directly


@pytest.mark.asyncio
async def test_rate_limit_proxy_gates_each_call_through_the_limiter() -> None:
    """RateLimitProxy awaits a token from the RateLimiter before each async call.

    Why this test is important:
        - The RateLimit layer (Job/Controller stacks, §6.3) must smooth bursts to the
          configured rate; a proxy that forgot to await the limiter would let a burst through.

    What it tests:
        - Two calls through the proxy each await the limiter once, and both reach the target.
    """

    class _Target:
        def __init__(self) -> None:
            self.calls = 0

        async def do(self) -> str:
            self.calls += 1
            return "ok"

    limiter = create_autospec(RateLimiter, instance=True)
    limiter.allow.return_value = True
    target = _Target()
    proxied = RateLimitProxy(target, limiter)

    assert await proxied.do() == "ok"
    assert await proxied.do() == "ok"
    assert limiter.wait.await_count == 2, "each call must await a rate-limit token"
    assert target.calls == 2

File: go/tests/unit/clients_client_stack_test.go
package unit_test

import (
	"context"
	"errors"
	"sync"
	"sync/atomic"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	decorators "github.com/gt-tech-ai/knowledge-engine/go/clients/decorators"
	"github.com/gt-tech-ai/knowledge-engine/go/foundation/resilience/bulkhead"
	"github.com/gt-tech-ai/knowledge-engine/go/foundation/resilience/circuitbreaker"
	"github.com/gt-tech-ai/knowledge-engine/go/foundation/resilience/retry"
)

// errBoom is a transient failure used to drive retry / circuit-breaker paths.
var errBoom = errors.New("boom")

// fastRetrier builds a real Retrier with tiny intervals so retry tests run in
// milliseconds rather than seconds.
func fastRetrier(t *testing.T, maxRetries int) *decorators.Stack {
	t.Helper()
	r, err := retry.NewFromConfig(retry.Config{
		Kind:            retry.KindExponential,
		MaxRetries:      maxRetries,
		InitialInterval: time.Millisecond,
		MaxInterval:     2 * time.Millisecond,
		Multiplier:      2.0,
		MaxElapsedTime:  time.Second,
	})
	require.NoError(t, err)
	return decorators.New("test").WithRetrier(r)
}

// TestClientStack_RetriesTransientThenSucceeds tests that a Retryable operation
// which fails transiently then succeeds is retried by the stack's Retry layer.
//
// Why this test is important:
//   - The widest-blast-radius direct-SDK calls must survive transient blips; the
//     Retry layer is the mechanism, and a Retryable:false op must NOT be retried.
//
// What it tests:
//   - A fn failing once then succeeding is invoked twice and returns success when
//     RunOpts.Retryable is true.
func TestClientStack_RetriesTransientThenSucceeds(t *testing.T) {
	t.Parallel()

	s := fastRetrier(t, 3)
	var calls int32
	out, err := decorators.Run(context.Background(), s, "get",
		decorators.RunOpts{Retryable: true},
		func(context.Context) (string, error) {
			if atomic.AddInt32(&calls, 1) == 1 {
				return "", errBoom
			}
			return "ok", nil
		})

	require.NoError(t, err)
	assert.Equal(t, "ok", out)
	assert.Equal(t, int32(2), atomic.LoadInt32(&calls), "should retry once then succeed")
}

// TestClientStack_NonRetryableIsNotRetried tests that a Retryable:false op is
// invoked exactly once even when it fails, so body-consuming ops (S3 Upload) are
// never replayed.
//
// Why this test is important:
//   - Replaying a body-consuming operation corrupts data; the per-op Retryable
//     flag is the guard.
//
// What it tests:
//   - A failing fn with RunOpts.Retryable false is invoked exactly once.
func TestClientStack_NonRetryableIsNotRetried(t *testing.T) {
	t.Parallel()

	s := fastRetrier(t, 3)
	var calls int32
	_, err := decorators.Run(context.Background(), s, "upload",
		decorators.RunOpts{Retryable: false},
		func(context.Context) (string, error) {
			atomic.AddInt32(&calls, 1)
			return "", errBoom
		})

	require.Error(t, err)
	assert.Equal(t, int32(1), atomic.LoadInt32(&calls), "non-retryable op must run once")
}

// TestClientStack_CircuitOpensAfterThreshold tests that the circuit breaker trips
// after repeated failures and then fails fast without invoking the wrapped fn.
//
// Why this test is important:
//   - Failing fast when a downstream is down is the point of the breaker; without
//     it a dead dependency ties up every caller.
//
// What it tests:
//   - After the failure threshold is crossed, a subsequent call returns an error
//     without invoking fn.
func TestClientStack_CircuitOpensAfterThreshold(t *testing.T) {
	t.Parallel()

	cb, err := circuitbreaker.New(circuitbreaker.KindGoBreaker, "test-cb",
		circuitbreaker.WithMaxRequests(1))
	require.NoError(t, err)
	s := decorators.New("test").WithCircuitBreaker(cb)

	// Trip the breaker: gobreaker's default ReadyToTrip opens after consecutive
	// failures; drive enough failing calls to open it.
	for i := 0; i < 10; i++ {
		_, _ = decorators.Run(context.Background(), s, "op", decorators.RunOpts{},
			func(context.Context) (string, error) { return "", errBoom })
	}

	var invoked bool
	_, err = decorators.Run(context.Background(), s, "op", decorators.RunOpts{},
		func(context.Context) (string, error) { invoked = true; return "ok", nil })

	require.Error(t, err, "open breaker should fail fast")
	assert.False(t, invoked, "open breaker must not invoke fn")
}

// TestClientStack_RetryAttemptsEachHitCircuitBreaker tests the load-bearing
// nesting order: Retry is OUTSIDE the CircuitBreaker, so each retry attempt is
// observed by the breaker, and once it opens mid-retry the remaining attempts
// fail fast without re-invoking fn.
//
// Why this test is important:
//   - If CB wrapped the whole retry loop instead of each attempt, the breaker
//     would see one aggregate outcome and never shed load between attempts —
//     inverting the prescribed Bulkhead→Retry→CircuitBreaker order.
//
// What it tests:
//   - With a low CB threshold and a higher retry budget, an always-failing fn is
//     invoked only up to the threshold (fewer than MaxRetries+1), proving the CB
//     short-circuits attempts inside the retry loop.
func TestClientStack_RetryAttemptsEachHitCircuitBreaker(t *testing.T) {
	t.Parallel()

	r, err := retry.NewFromConfig(retry.Config{
		Kind: retry.KindExponential, MaxRetries: 8,
		InitialInterval: time.Millisecond, MaxInterval: 2 * time.Millisecond,
		Multiplier: 2.0, MaxElapsedTime: time.Second,
	})
	require.NoError(t, err)
	cb, err := circuitbreaker.New(circuitbreaker.KindGoBreaker, "nest-cb")
	require.NoError(t, err)
	s := decorators.New("test").WithRetrier(r).WithCircuitBreaker(cb)

	var calls int32
	_, err = decorators.Run(context.Background(), s, "op",
		decorators.RunOpts{Retryable: true},
		func(context.Context) (string, error) {
			atomic.AddInt32(&calls, 1)
			return "", errBoom
		})

	require.Error(t, err)
	// gobreaker's default ReadyToTrip opens after >5 consecutive failures, so fn is
	// invoked ~6 times before the breaker short-circuits the remaining retries. A snug
	// upper bound (7) still catches the inverted-nesting regression (which would invoke
	// fn all 9 attempts) AND a subtler one where the CB opens several attempts too late.
	assert.LessOrEqual(
		t,
		atomic.LoadInt32(&calls),
		int32(7),
		"CB should open after ~6 failures and stop re-invoking fn well before MaxRetries+1",
	)
}

// TestClientStack_BulkheadLimitsConcurrency tests that the bulkhead bounds the
// number of concurrently in-flight operations.
//
// Why this test is important:
//   - Unbounded parallelism against a slow dependency exhausts goroutines/FDs;
//     the bulkhead is the cap.
//
// What it tests:
//   - With MaxConcurrent=1, two concurrent Runs never overlap (peak in-flight 1).
func TestClientStack_BulkheadLimitsConcurrency(t *testing.T) {
	t.Parallel()

	bh, err := bulkhead.New(bulkhead.KindChannel, bulkhead.WithMaxConcurrent(1))
	require.NoError(t, err)
	s := decorators.New("test").WithBulkhead(bh)

	var inFlight, peak int32
	var wg sync.WaitGroup
	op := func(context.Context) (string, error) {
		n := atomic.AddInt32(&inFlight, 1)
		for {
			p := atomic.LoadInt32(&peak)
			if n <= p || atomic.CompareAndSwapInt32(&peak, p, n) {
				break
			}
		}
		time.Sleep(10 * time.Millisecond)
		atomic.AddInt32(&inFlight, -1)
		return "ok", nil
	}
	for i := 0; i < 2; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			_, _ = decorators.Run(context.Background(), s, "op", decorators.RunOpts{}, op)
		}()
	}
	wg.Wait()

	assert.Equal(
		t,
		int32(1),
		atomic.LoadInt32(&peak),
		"bulkhead must cap concurrency at 1",
	)
}

// TestClientStack_PreservesConfiguredTimeout tests that the Timeout layer applies
// the configured deadline verbatim and never imposes a shorter default.
//
// Why this test is important:
//   - A deliberate long timeout (e.g. Ollama's 300s embed) must not be silently
//     shortened by the stack — the AC forbids it.
//
// What it tests:
//   - The fn observes a context deadline within tolerance of the configured
//     timeout; a stack with zero timeout imposes no deadline.
func TestClientStack_PreservesConfiguredTimeout(t *testing.T) {
	t.Parallel()

	// Configured long timeout is passed through, not shortened.
	s := decorators.New("test").WithTimeout(90 * time.Second)
	_, err := decorators.Run(context.Background(), s, "op", decorators.RunOpts{},
		func(ctx context.Context) (string, error) {
			dl, ok := ctx.Deadline()
			require.True(t, ok, "timeout layer should set a deadline")
			assert.InDelta(
				t,
				90*time.Second,
				time.Until(dl),
				float64(2*time.Second),
				"deadline must match configured timeout",
			)
			return "ok", nil
		})
	require.NoError(t, err)

	// Zero timeout imposes no deadline.
	s0 := decorators.New("test")
	_, err = decorators.Run(context.Background(), s0, "op", decorators.RunOpts{},
		func(ctx context.Context) (string, error) {
			_, ok := ctx.Deadline()
			assert.False(t, ok, "zero timeout must not impose a deadline")
			return "ok", nil
		})
	require.NoError(t, err)
}

// TestClientStack_DisabledConfigIsPassthrough tests that a disabled Config yields
// a bare stack that invokes fn directly with no resilience layers.
//
// Why this test is important:
//   - The AC requires a backwards-compatible path: config off → bare client.
//
// What it tests:
//   - StackFromConfig(Enabled:false) returns a stack that runs fn once with no
//     imposed deadline (no timeout layer) and returns its result verbatim.
func TestClientStack_DisabledConfigIsPassthrough(t *testing.T) {
	t.Parallel()

	s, err := decorators.StackFromConfig(
		"test",
		decorators.Config{Enabled: false},
		decorators.Deps{},
	)
	require.NoError(t, err)

	var calls int32
	out, err := decorators.Run(
		context.Background(),
		s,
		"op",
		decorators.RunOpts{Retryable: true},
		func(ctx context.Context) (string, error) {
			atomic.AddInt32(&calls, 1)
			_, ok := ctx.Deadline()
			assert.False(t, ok, "disabled stack imposes no timeout")
			return "bare", nil
		},
	)

	require.NoError(t, err)
	assert.Equal(t, "bare", out)
	assert.Equal(t, int32(1), atomic.LoadInt32(&calls))
}
